### PR TITLE
Drop async IO

### PR DESCRIPTION
**Pull request by dpc**
For a relatively simple functionality like this
deamon, blocking IO + threads are sufficient.

### CODE REVIEW

**Comment by dpc** [comment]
I added some Rust integration tests, because it's just easier to run, and I've found inttest.sh afterwards. I guess having a two different tests is even more confidence that nothing broke.

BTW. [Personal AI usage disclosure](https://dpc.pw/posts/personal-ai-usage-disclosure/). I wrote projects like that before, and everything is fairly straightforward. I did review the code.

File: src/lib.rs [hunk 1]
```diff
@@ -66,72 +69,105 @@ fn create_listener(socket_path: &Path) -> Result<UnixListener> {
...
+    });
+
     debug!("Entering main loop");
-    let mut stop = None;
-    while stop.is_none() {
-        select! {
-            result = listener.accept() => match result {
-                Ok((socket, _addr)) => {
-                    debug!("Connection accepted");
-                    // TODO(jmmv): Connections are handled sequentially.  This is... fine.
```
Comment (dpc) [inline src/lib.rs:116]: Might as well take care of this one too.

File: src/lib.rs [hunk 2]
```diff
@@ -66,72 +69,105 @@ fn create_listener(socket_path: &Path) -> Result<UnixListener> {
...
         None => {
-            return Err("No agent found; cannot proxy request".to_owned());
+            warn!("Dropping connection: no agent found");
+            return;
         }
     };
-    let result = tokio::io::copy_bidirectional(&mut client, &mut agent)
-        .await.map(|_| ())
-        .map_err(|e| format!("{}", e));
+    if let Err(e) = handle_bi_socket_forwarding(client, agent) {
```
Comment (jmmv) [inline src/lib.rs:103]: Before I switched to tokio's `copy_bidirectional`, I had actually written my own proxying logic but then dropped it once an AI review told me tokio had a built-in for this.

I'd want to add some unit tests for this for various corner cases -- and I had written them in that previous iteration. See https://github.com/jmmv/ssh-agent-switcher/commit/adaa31fc0b6982a19825e3d8388ffd941af93435 . Maybe they can be reused.

File: src/lib.rs [hunk 3]
```diff
@@ -66,72 +69,105 @@ fn create_listener(socket_path: &Path) -> Result<UnixListener> {
...
+    let agents_dirs: Arc<[PathBuf]> = agents_dirs.into();
 
     let listener = create_listener(&socket_path)?;
 
+    // Set up signal handling with the atomic flag + reconnect pattern
+    let shutdown = Arc::new(AtomicBool::new(false));
+    let shutdown_clone = Arc::clone(&shutdown);
+    let socket_path_clone = socket_path.clone();
+
+    let mut signals = Signals::new([SIGHUP, SIGINT, SIGQUIT, SIGTERM])
```
Comment (jmmv) [inline src/lib.rs:126]: I had tried this before switching to async, but then encountered issues because signal-hook sets `SA_RESTART` unconditionally and that made it "impossible" to reliably avoid race conditions. I then encountered https://github.com/vorner/signal-hook/issues/153 and all of the discussion there discouraged me from trying to get it right, and switching to async made the code much clearer / easier to follow. Are you confident that those issues don't exist here?

File: src/lib.rs (hunk 2 above)
Comment (dpc) [reply src/lib.rs:103]: I specifically use two threads to handle both directions each to avoid having to deal with bidrection. The logic of every direction is trivial and can't really go too wrong. For a low usage daemon like this one, an extra thread is negligible cost.

Adding more corner case tests is a good idea anyway. Will do.

File: src/lib.rs (hunk 3 above)
Comment (dpc) [reply src/lib.rs:126]: This method specifically avoids relying on any io interruption behavior. This thread is standalone, and when it receives a signal will make fake connection to the listener. Unless I'm missing something, the mentioned issue does not apply. Not every application is so lucky though to have a good way to wake up the main loop reliably, so yeah, this could be an issue. But if I had to start working it around, I would just use async Rust. :shrug:

File: src/lib.rs (hunk 2 above)
Comment (dpc) [reply src/lib.rs:103]: 9b143c0e - LLM generated, but I don't think it matters much, it's just a sanity check. Please let me know if you have any other scenarios in mind.

File: src/lib.rs (hunk 3 above)
Comment (jmmv) [reply src/lib.rs:126]: Yeah, I think when I looked at this a month ago I was trying to use just one extra thread, not two, and then it's when I ran into these issues.

**Comment by dpc** [comment]
I am no longer in favor of this approach. It works OK in selfci, because the sockets are relatively static, but once I started playing with this and systemfd activation, it became apparent that calling yourself through unix path can sometimes lead to connecting to something competely different that expected, or not work at all (because there is no socket path).
